prow/tide/contextregister.go

```diff
@@ -0,0 +1,145 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tide
+
+import (
+	"sync"
+
+	"github.com/shurcooL/githubql"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/test-infra/prow/config"
+)
+
+type contextChecker interface {
+	// ignoreContext tells whether a context is optional.
+	ignoreContext(context Context) bool
+	// missingRequiredContexts tells if required contexts are missing from the list of contexts provided.
+	missingRequiredContexts([]Context) []Context
+}
+
+// newExpectedContext creates a Context with Expected state.
+// This should not only be used when contexts are missing.
+func newExpectedContext(c string) Context {
+	return Context{
+		Context:     githubql.String(c),
+		State:       githubql.StatusStateExpected,
+		Description: githubql.String(""),
+	}
+}
+
+// contextRegister implements contextChecker and allow registering of required and optional contexts.
+type contextRegister struct {
+	lock                sync.RWMutex
+	required, optional  sets.String
+	skipUnknownContexts bool
+}
+
+// newContextRegister instantiates a new contextRegister and register tide as optional by default
+// and uses Prow Config to find optional and required tests, as well as skipUnknownContexts.
+func newContextRegister(skipUnknownContexts bool) *contextRegister {
+	r := &contextRegister{
+		required:            sets.NewString(),
+		optional:            sets.NewString(),
+		skipUnknownContexts: skipUnknownContexts,
+	}
+	r.registerOptionalContexts(statusContext)
+	return r
+}
+
+// newContextRegisterFromPolicy instantiates a new contextRegister and register tide as optional by default
+// and uses Prow Config to find optional and required tests, as well as skipUnknownContexts.
+func newContextRegisterFromPolicy(policy *config.Policy) *contextRegister {
+	r := newContextRegister(false)
+	if policy != nil {
+		if policy.SkipUnknownContexts != nil {
+			r.skipUnknownContexts = *policy.SkipUnknownContexts
+		}
+		if policy.Protect != nil && *policy.Protect {
+			r.registerRequiredContexts(policy.Contexts...)
+		} else if r.skipUnknownContexts {
```

Comment:
If the policy is not to protect but we are skipping unknown contexts we want to require the contexts the policy has listed to not protect on? What?